### ISSUE TITLE
Set up container registry and CI/CD deployment pipeline

### ISSUE THREAD

**Issue by omniscient** [omniscient/markethawk#104]
## Context

From the [Architecture & Quality Report](./architecture-quality-report.html) — Deployment Readiness scored 3/5.

Docker images are built locally on each `docker-compose up --build`. There is no container registry, no image versioning, and no automated deployment pipeline. Deployments are manual and not reproducible — the same code could produce different images depending on when `pip install` and `npm install` run.

## Evidence

- No Docker Hub, GHCR, or private registry referenced in `docker-compose.yml`
- All services use `build: context: ./backend` or `build: context: ./frontend` — local builds only
- No image tagging strategy (no `:latest`, no git SHA tags)
- No deployment automation (no GitHub Actions deploy workflow)
- CI workflow only runs tests — no image build step

## Scope

1. **Set up GitHub Container Registry (GHCR)**:
   - Configure `docker/login-action` in CI
   - Build and push backend, frontend, and dark-factory images
   - Tag with git SHA and `latest`
2. **Add image build to CI**:
   - Build images on PR merge to main
   - Run security scan on built images (e.g., `trivy`)
3. **Update `docker-compose.yml`** to support both local build and registry pull:
   ```yaml
   backend:
     image: ghcr.io/omniscient/markethawk-backend:latest
     build:
       context: ./backend
   ```
4. **Create deployment workflow** (manual trigger or tag-based):
   - Pull latest images
   - Run migrations
   - Rolling restart services
5. **Document rollback procedure**: `docker-compose pull` a specific tag

## Benefit

Reproducible deployments, rollback capability, image security scanning, faster deploys (pull vs. build).

## Effort: L

---
*From Architecture & Quality Report — Section 14: Prioritized Improvement Roadmap (Strategic, 6+ months)*

**Comment by omniscient** [comment]
🧠 **Refinement Pipeline** — Starting brainstorming and spec generation.

---
*Posted by MarketHawk Backlog Scheduler*

**Comment by omniscient** [comment]
## Refinement Pipeline — Spec Generated

**Spec:** [Docs/superpowers/specs/2026-05-28-container-registry-cicd-design.md](https://github.com/omniscient/markethawk/blob/refine/issue-104-set-up-container-registry-and-ci-cd-depl/Docs/superpowers/specs/2026-05-28-container-registry-cicd-design.md)
**Branch:** [`refine/issue-104-set-up-container-registry-and-ci-cd-depl`](https://github.com/omniscient/markethawk/tree/refine/issue-104-set-up-container-registry-and-ci-cd-depl)

### Summary

The spec covers publishing three Docker images (backend, frontend, dark-factory) to GHCR on every merge to `main`, tagged with git SHA and `latest`. A separate `workflow_dispatch` deployment workflow SSH-deploys to a server, pulling images and running migrations — scaffolded with placeholder secrets that activate once a server is provisioned. Trivy security scanning is informational only (SARIF uploaded to GitHub Security tab), and `docker-compose.yml` is updated to support both local builds and registry pulls simultaneously.

### Brainstorming Q&A

> **Q1:** The docker-compose.yml has five distinct build contexts: `./backend/Dockerfile` (shared by backend/celery-worker/celery-beat/live-scanner/flower), `./frontend/Dockerfile`, `dark-factory/Dockerfile`, `./services/tweet-monitor/Dockerfile`, and `./backend/Dockerfile.forecast`. The issue explicitly names backend, frontend, and dark-factory. Should tweet-monitor and forecast-worker also be published to GHCR?
>
> **A1:** Publish three images: `markethawk-backend`, `markethawk-frontend`, `markethawk-dark-factory`. Exclude tweet-monitor (architecturally independent, separate cadence) and forecast-worker (profile-based, requires 36GB RAM, opt-in feature). Both can be added in follow-up issues.

> **Q2:** Vite bakes `VITE_*` env vars into the JavaScript bundle at build time — they cannot be changed at container runtime. The current docker-compose.yml sets `VITE_API_TARGET: http://backend:8000`. For the published GHCR image, should this be a build arg (environment-specific image), a runtime injection pattern (window.API_TARGET), or just bake the Docker network URL as-is?
>
> **A2:** Bake `http://backend:8000` as-is. The frontend Dockerfile runs in dev mode (Vite dev server proxy). All deployments use Docker Compose with the backend container named `backend` on the same Docker bridge network, so the URL is always correct. No code changes needed.

> **Q3:** The issue says "manual trigger or tag-based" for the deployment workflow. Should this be `workflow_dispatch` only, tag-push triggered, or both?
>
> **A3:** `workflow_dispatch` only. The deployment server's hostname, SSH keys, and access method are not yet documented anywhere in the codebase. A manual trigger with parameterized tag input can be scaffolded now and activated when server details are configured. Tag-based automation adds complexity without clear benefit at current deployment maturity.

> **Q4:** Trivy security scanning is mentioned in the issue. Should failing CRITICAL/HIGH CVEs block the image push (hard gate), or should results be uploaded to GitHub Security tab without failing the workflow (informational)?
>
> **A4:** Informational only. Both backend (Python + scikit-learn + LightGBM) and frontend (Node 22 + 30+ npm packages) have deep dependency trees with likely transitive CVEs. A hard gate would block all deploys until upstream vendors patch transitive dependencies. The team has already done a CVE cleanup round (commit `9e9da92`). Informational scanning gives visibility without blocking releases.

### Requirements

- Build and push `markethawk-backend`, `markethawk-frontend`, `markethawk-dark-factory` to GHCR on every push to `main`
- Tag each image with `sha-{GIT_SHA}` (permanent) and `latest` (moves forward)
- All backend-derived services (celery-worker, celery-beat, live-scanner, flower) reference the same `markethawk-backend` image
- `docker-compose.yml` updated with dual `image:` + `build:` fields so local builds and registry pulls both work
- Trivy scans all three images and uploads SARIF to GitHub Security tab; non-blocking
- `workflow_dispatch` deployment workflow with `image_tag` input (default: `latest`) and `run_migrations` toggle
- Rollback by pinning `image:` to a prior `sha-{SHA}` tag and re-deploying

### Approach

Two new GitHub Actions workflows: `ci-publish.yml` (triggers on push to `main`, builds three images in parallel, runs Trivy) and `deploy.yml` (manual trigger, SSH to server, `docker compose pull` + `up -d` + migration). The existing `ci.yml` test workflow is unchanged.

### Assumptions

- `GITHUB_TOKEN` has `packages: write` permission (standard for GHCR in the same repo)
- Deployment server runs Docker Compose v2 (`docker compose`)
- The existing Dockerfiles are dev-mode images and are published as-is (production multi-stage builds are out of scope)
- `dark-factory/Dockerfile` build context is the full repo root (`.`)

### Next Steps

- ✅ **Approve spec** — move the issue to the **Refined** column on the project board. The scheduler will automatically trigger plan generation.
- ✏️ **Request changes** — leave a comment on this issue with your feedback, then re-run:
  ```bash
  docker compose --profile factory run --rm dark-factory "Refine issue #104"
  ```
- ❓ **Need to discuss** — add the `needs-discussion` label to pause automation.

---
*Posted by MarketHawk Refinement Pipeline*

**Comment by omniscient** [comment]
📋 **Refinement Pipeline** — Starting plan generation and architect validation.

---
*Posted by MarketHawk Backlog Scheduler*

**Comment by omniscient** [comment]
## Refinement Pipeline — Plan Generated

**Plan:** [`Docs/superpowers/plans/2026-05-29-container-registry-cicd.md`](https://github.com/omniscient/markethawk/blob/refine/issue-104-set-up-container-registry-and-ci-cd-depl/Docs/superpowers/plans/2026-05-29-container-registry-cicd.md)
**Branch:** [`refine/issue-104-set-up-container-registry-and-ci-cd-depl`](https://github.com/omniscient/markethawk/tree/refine/issue-104-set-up-container-registry-and-ci-cd-depl)
**Tasks:** 4 tasks, 19 steps

### Task Overview

1. **Create `ci-publish.yml`** — Parallel GHCR image builds for backend, frontend, and dark-factory on every push to `main`; Trivy security scanning (non-blocking, SARIF to GitHub Security tab)
2. **Create `deploy.yml`** — Scaffolded `workflow_dispatch` SSH deployment workflow; `image_tag` input (default: `latest`) lets operators pin to any SHA tag; requires `DEPLOY_HOST`/`DEPLOY_USER`/`DEPLOY_SSH_KEY` secrets once a server is provisioned
3. **Update `docker-compose.yml`** — Add `image: ghcr.io/omniscient/markethawk-{service}:${IMAGE_TAG:-latest}` to 8 built services, preserving `build:` for local development; env-var tag pattern enables pinned deployments without mutating the compose file
4. **Document rollback procedure** — Append SHA tag pinning guide to `deployment-guide.md`

### Architect Review

> **Cycle 1:**
> **Verdict:** Issues Found
> **Feedback:** Plan summary passed to the architect omitted the full YAML blocks; also flagged that using `sed` to pin image tags in `deploy.yml` is fragile.
> **Changes made:** Replaced `sed` with `${IMAGE_TAG:-latest}` env var substitution in all docker-compose image fields. All 8 services use the env-var pattern so the compose file never needs mutation on the server — exporting `IMAGE_TAG=sha-abc1234` before `docker compose pull` is sufficient.

> **Cycle 2:**
> **Verdict:** Approved
> **Feedback:** Approved with advisory notes: `flower` and `backlog-scheduler` were missing from the `deploy.yml` restart commands (inconsistency with R3), `envs: IMAGE_TAG` was redundant alongside inline substitution, and a comment about profile-gated services (`dark-factory`, `backlog-scheduler`) was suggested.
> **Changes made:** Added `flower` and `backlog-scheduler` to the `docker compose pull` and `up` commands; removed redundant `envs:` line; added inline comment noting that `dark-factory` is profile-gated and updated via pull.

### Next Steps

- ✅ **Approve plan** — move the issue to the **Ready** column on the project board. The scheduler will automatically start implementation.
- ✏️ **Request changes** — leave a comment on this issue with your feedback, then re-run:
  ```bash
  docker compose --profile factory run --rm dark-factory "Plan issue #104"
  ```
- ❓ **Need to discuss** — add the `needs-discussion` label to pause automation.

---
*Posted by MarketHawk Refinement Pipeline*

**Comment by omniscient** [comment]
## Dark Factory — Cost Report

**2 run(s) — Total: $3.15709630 (102 in / 19.5K out)**

### Run: 2026-05-29 21:43 UTC (plan, completed)

| Step | Model | In tokens | Out tokens | Cost | Duration |
|------|-------|-----------|------------|------|----------|
| parse-intent |  | 28 | 298 | $0.0162 | 4.6s |
| fetch-issue |  | 0 | 0 | $0 | 960ms |
| setup-refine-branch |  | 0 | 0 | $0 | 430ms |
| plan |  | 11 | 4K | $1.6297 | 10m 20s |
| plan-push-and-advance |  | 0 | 0 | $0 | 287ms |
| **Subtotal** | | **39** | **4.2K** | **$1.64593325** | |
### Run: 2026-06-02 02:59 UTC (fix, completed)

| Step | Model | In tokens | Out tokens | Cost | Duration |
|------|-------|-----------|------------|------|----------|
| parse-intent |  | 28 | 325 | $0.0164 | 4.8s |
| fetch-issue |  | 0 | 0 | $0 | 928ms |
| setup-branch |  | 0 | 0 | $0 | 32ms |
| implement |  | 32 | 14.8K | $0.8568 | 4m 18s |
| preview-up |  | 0 | 0 | $0 | 23.1s |
| validate |  | 3 | 144 | $0.6379 | 2m 49s |
| push-and-pr |  | 0 | 0 | $0 | 2.9s |
| status-in-review |  | 0 | 0 | $0 | 1.7s |
| report |  | 0 | 0 | $0 | 1.1s |
| **Subtotal** | | **63** | **15.2K** | **$1.51116305** | |

---
*Updated by MarketHawk Dark Factory*

**Comment by omniscient** [comment]
Related to #146 (local dev bind-mounts the working tree into the running containers, so editing/branch-switching mutates the live stack). #146 is the focused compose-layer slice — split dev (bind-mount + `--reload`) from base (run the built image, no bind-mount) — and is a **prerequisite for option 3 here** (serving registry/built images instead of the working tree). Keeping #146 as a near-term standalone fix that this epic subsumes.

**Comment by omniscient** [comment]
## Dark Factory Run — New implementation

**PR:** #155

**Branch:** `feat/issue-104-set-up-container-registry-and-ci-cd-depl`

### Changes

# Implementation Summary — Issue #104

## Files Created

### `.github/workflows/ci-publish.yml`
- Triggers on push to `main`
- 3 parallel build jobs: `build-backend`, `build-frontend`, `build-dark-factory`
- Each job: GHCR login → metadata generation (sha-{SHORT_SHA} + latest tags) → build-push
- 1 scan job (needs all builds): Trivy matrix scan over 3 images, SARIF uploaded to GitHub Security tab
- `continue-on-error: true` on scan job (informational, non-blocking)
- Permissions: `packages: write` + `security-events: write`

### `.github/workflows/deploy.yml`
- `workflow_dispatch` trigger with two inputs:
  - `image_tag` (string, default: `latest`) — pin to any SHA tag for rollback
  - `run_migrations` (boolean, default: `true`)
- SSH deployment via `appleboy/ssh-action@v1`
- Pulls: backend, celery-worker, celery-beat, live-scanner, flower, frontend, backlog-scheduler
- Also pulls dark-factory (profile=factory — updated but not auto-restarted)
- Restarts all main services + backlog-scheduler (--profile scheduler)
- Conditionally runs Alembic migrations
- Scaffolded with placeholder secrets (DEPLOY_HOST, DEPLOY_USER, DEPLOY_SSH_KEY)

## Files Modified

### `docker-compose.yml`
- Added `image: ghcr.io/omniscient/markethawk-backend:${IMAGE_TAG:-latest}` to:
  `backend`, `celery-worker`, `celery-beat`, `live-scanner`, `flower`
- Added `image: ghcr.io/omniscient/markethawk-frontend:${IMAGE_TAG:-latest}` to:
  `frontend`
- Added `image: ghcr.io/omniscient/markethawk-dark-factory:${IMAGE_TAG:-latest}` to:
  `dark-factory`, `backlog-scheduler`
- All `build:` fields preserved — local `docker compose up --build` still works

### `deployment-guide.md`
- Added "Deploying from GHCR" section: describes sha + latest tag strategy, IMAGE_TAG usage
- Added "Rollback Procedure" section: SHA tag pinning with GHCR package browser link and alembic downgrade reminder

## Tests Added
None — this is purely CI/CD configuration (YAML files + documentation). YAML syntax was validated with Python's `yaml.safe_load`. No backend Python changes, so the existing pytest suite is unaffected.

## Migrations Created
None — no SQLAlchemy model changes.

## Decisions and Trade-offs

1. **Three images only** (backend, frontend, dark-factory): tweet-monitor and forecast-worker excluded per spec Q&A — architecturally independent cadences.

2. **Trivy informational-only** (`exit-code: 0`): both images have deep transitive dependency trees. Hard gate would block all merges until upstream vendors patch CVEs. SARIF upload gives visibility in GitHub Security tab without blocking.

3. **workflow_dispatch only** (not tag-push): deployment server details not yet documented. Manual trigger activates once DEPLOY_HOST/DEPLOY_USER/DEPLOY_SSH_KEY secrets are set.

4. **`${IMAGE_TAG:-latest}` pattern**: compose file never needs mutation on the server — exporting `IMAGE_TAG=sha-abc1234` before docker compose commands is sufficient for rollback.

5. **Existing `ci.yml` unchanged**: test workflow remains on PR triggers; publish workflow triggers only on push to main.

### Preview Environment

| Service | URL |
|---------|-----|
| Frontend | http://localhost:10533 |
| Backend API | http://mh-preview-104-backend-1:8000 |
| API Docs | http://mh-preview-104-backend-1:8000/docs |
| PostgreSQL | `localhost:10554` |
| Redis | `localhost:10563` |

### Commands
```bash
# Iterate after review
docker compose --profile factory run --rm dark-factory "Continue issue #104"

# Tear down when done
docker compose --profile factory run --rm dark-factory "Close issue #104"
```

---
*Posted by MarketHawk Dark Factory*
